File: backend/entities/Building.cs
using System;
using System.Collections.Generic;

namespace Pims.Dal.Entities
{
    /// <summary>
    /// Building class, provides an entity for the datamodel to manage buildings.
    /// </summary>
    public class Building : Property
    {
        #region Properties
        /// <summary>
        /// get/set - The foreign key to the property building construction type.
        /// </summary>
        public int BuildingConstructionTypeId { get; set; }

        /// <summary>
        /// get/set - The building construction type for this property.
        /// </summary>
        public BuildingConstructionType BuildingConstructionType { get; set; }

        /// <summary>
        /// get/set - The number of floors in the building.
        /// </summary>
        public int BuildingFloorCount { get; set; }

        /// <summary>
        /// get/set - The foreign key to the building predominant use.
        /// </summary>
        public int BuildingPredominateUseId { get; set; }

        /// <summary>
        /// get/set - The building predominant use for this building.
        /// </summary>
        public BuildingPredominateUse BuildingPredominateUse { get; set; }

        /// <summary>
        /// get/set - The type of tenancy for this building.
        /// </summary>
        public string BuildingTenancy { get; set; }

        /// <summary>
        /// get/set - The date the building tenancy was last updated.
        /// </summary>
        public DateTime? BuildingTenancyUpdatedOn { get; set; }

        /// <summary>
        /// get/set - The building rentable area.
        /// </summary>
        public float RentableArea { get; set; }

        /// <summary>
        /// get/set - The building total area.
        /// </summary>
        public float TotalArea { get; set; }

        /// <summary>
        /// get/set - The foreign key to the building occupant type.
        /// </summary>
        public int BuildingOccupantTypeId { get; set; }

        /// <summary>
        /// get/set - The type of occupant for this building.
        /// </summary>
        public BuildingOccupantType BuildingOccupantType { get; set; }

        /// <summary>
        /// get/set - The expiry date of the currently active lease
        /// </summary>
        public DateTime? LeaseExpiry { get; set; }

        /// <summary>
        /// get/set - The name of the occupant/organization
        /// </summary>
        public string OccupantName { get; set; }

        /// <summary>
        /// get/set - Whether the lease on this building would be transferred if the building is sold.
        /// </summary>
        public bool TransferLeaseOnSale { get; set; } = false;

        /// <summary>
        /// get/set - Metadata related to the buildings leased status.
        /// </summary>
        public string LeasedLandMetadata { get; set; }

        /// <summary>
        /// get - A collection of parcels this building is located on.
        /// </summary>
        public ICollection<ParcelBuilding> Parcels { get; } = new List<ParcelBuilding>();

        /// <summary>
        /// get - A collection of evaluations for this building.
        /// </summary>
        /// <typeparam name="BuildingEvaluation"></typeparam>
        public ICollection<BuildingEvaluation> Evaluations { get; } = new List<BuildingEvaluation>();

        /// <summary>
        /// get - A collection of fiscal values for this building.
        /// </summary>
        /// <typeparam name="BuildingFiscals"></typeparam>
        public ICollection<BuildingFiscal> Fiscals { get; } = new List<BuildingFiscal>();

        /// <summary>
        /// get - A collection of projects this building is assocated to.
        /// </summary>
        public ICollection<ProjectProperty> Projects { get; } = new List<ProjectProperty>();
        #endregion

        #region Constructors
        /// <summary>
        /// Create a new instance of a Building class.
        /// </summary>
        public Building() { }

        /// <summary>
        /// Create a new instance of a Building class.
        /// </summary>
        /// <param name="parcel"></param>
        /// <param name="latitude"></param>
        /// <param name="longitude"></param>
        public Building(Parcel parcel, double latitude, double longitude) : base(latitude, longitude)
        {
            if (parcel != null) {
                var pb = new ParcelBuilding(parcel, this);
                this.Parcels.Add(pb);
            }
        }
        #endregion
    }
}

File: express-api/src/typeorm/Entities/Building.ts
import { BuildingConstructionType } from '@/typeorm/Entities/BuildingConstructionType';
import { BuildingOccupantType } from '@/typeorm/Entities/BuildingOccupantType';
import { BuildingPredominateUse } from '@/typeorm/Entities/BuildingPredominateUse';
import { Entity, Column, ManyToOne, Index, JoinColumn, OneToMany } from 'typeorm';
import { Property } from '@/typeorm/Entities/abstractEntities/Property';
import { BuildingFiscal } from '@/typeorm/Entities/BuildingFiscal';
import { BuildingEvaluation } from '@/typeorm/Entities/BuildingEvaluation';

// Should the occupant information (OccupantTypeId, OccupantName, and BuildingTenancyUpdatedOn) be a separate table?
// This may change over time and we wouldn't be able to track previous occupants by storing it in this table.

export enum LeasedLandTypes {
  OWNED = 0,
  LEASED = 1,
  OTHER = 2,
}
export interface LeasedLandMetadata {
  Type?: LeasedLandTypes;
  ParcelId?: number;
  OwnershipNote?: string;
}

@Entity()
@Index(['PID', 'PIN'], { unique: false })
export class Building extends Property {
  // Construction Type Relations
  @Column({ name: 'building_construction_type_id', type: 'int' })
  BuildingConstructionTypeId: number;

  @ManyToOne(() => BuildingConstructionType, (ConstructionType) => ConstructionType.Id)
  @JoinColumn({ name: 'building_construction_type_id' })
  @Index()
  BuildingConstructionType: BuildingConstructionType;

  @Column({ type: 'int' })
  BuildingFloorCount: number;

  // Predominate Use Relations
  @Column({ name: 'building_predominate_use_id', type: 'int' })
  BuildingPredominateUseId: number;

  @ManyToOne(() => BuildingPredominateUse, (PredominateUse) => PredominateUse.Id)
  @JoinColumn({ name: 'building_predominate_use_id' })
  @Index()
  BuildingPredominateUse: BuildingPredominateUse;

  @Column({ type: 'character varying', length: 450 })
  BuildingTenancy: string;

  @Column({ type: 'real' })
  RentableArea: number;

  // Occupant Type Relations
  @Column({ name: 'building_occupant_type_id', type: 'int', nullable: true })
  BuildingOccupantTypeId: number;

  @ManyToOne(() => BuildingOccupantType, (OccupantType) => OccupantType.Id)
  @JoinColumn({ name: 'building_occupant_type_id' })
  @Index()
  BuildingOccupantType: BuildingOccupantType;

  @Column({ type: 'timestamp', nullable: true })
  BuildingTenancyUpdatedOn: Date;

  @Column({ type: 'real' })
  TotalArea: number;

  @OneToMany(() => BuildingFiscal, (Fiscal) => Fiscal.Building, { nullable: true, cascade: true })
  Fiscals: BuildingFiscal[];

  @OneToMany(() => BuildingEvaluation, (Evaluation) => Evaluation.Building, {
    nullable: true,
    cascade: true,
  })
  Evaluations: BuildingEvaluation[];
}
